internal/terraform/upgrade_resource_state.go

```diff
@@ -136,8 +137,12 @@ func upgradeResourceState(addr addrs.AbsResourceInstance, provider providers.Int
 // stripRemovedStateAttributes deletes any attributes no longer present in the
 // schema, so that the json can be correctly decoded.
 func stripRemovedStateAttributes(state []byte, ty cty.Type) []byte {
+	// we must use json.Number to avoid changing the precision of cty.Number values
+	decoder := json.NewDecoder(bytes.NewReader(state))
+	decoder.UseNumber()
+
 	jsonMap := map[string]interface{}{}
-	err := json.Unmarshal(state, &jsonMap)
+	err := decoder.Decode(&jsonMap)
```

Comment:
I don't know how relevant this is here, but there is a subtle difference in behaviour between `Unmarshal` and the decoder. 

Unmarshal will consume the whole input, and fail if the whole thing doesn't match valid JSON.

The decoder will more than happily partially consume the input as long as it starts with valid JSON, but then leave anything at the end that is no longer JSON.

For example: `{"key":"value"}nomorejson`. The unmarshal function would return an error for this, while the decoder will accept it and just drop the "nomorejson" part.

See this PR for an equivalent fix we had to make in the renderer: https://github.com/hashicorp/terraform/pull/34959

Reply:
oh yes I had forgotten about that issue! I don't think that should matter here however, because we are always passing in a json encoded state, so it must by definition always be a single valid json object. If we got a state like `{"key":"value"}nomorejson` something else has gone horribly wrong and it will fail during normal decoding anyway.